### PR TITLE
feat: pass event log to event triggering functions

### PR DESCRIPTION
**Pull request by mkykadir**

### CODE REVIEW

File: src/lib/types/Web3FunctionContext.ts [hunk 1]
```diff
@@ -1,6 +1,23 @@
 import { BigNumber } from "@ethersproject/bignumber";
 import { Web3FunctionMultiChainProvider } from "../provider/Web3FunctionMultiChainProvider";
 import { Web3FunctionUserArgs } from "./Web3FunctionUserArgs";
+
+export interface Log {
```
[the remark this answers is not in the document]
Comment (mkykadir) [reply src/lib/types/Web3FunctionContext.ts:5]: Yeah we can driectly use that but from the task I understood, have a separate interface wdyt @goums ?

[the remark this answers is not in the document]
Comment (goums) [reply src/lib/types/Web3FunctionContext.ts:5]: @mkykadir I was also thinking of re-using the type from `ethers`
we already use them for providers & BigNumbers

[the remark this answers is not in the document]
Comment (mkykadir) [reply src/lib/types/Web3FunctionContext.ts]: okay refactored in ed2af26b90c39dc9cd3eafb0e780c12fac946d6b

File: src/lib/Web3Function.ts [hunk 2]
```diff
@@ -121,7 +131,10 @@ export class Web3Function {
       },
     };
 
-    const result = await this._onRun(context);
+    const result = ctxData.log
+      ? await (this._onRun as eventRunHandler)({ ...context, log: ctxData.log })
+      : await (this._onRun as baseRunHandler)(context);
+
```
Comment (goums) [inline src/lib/Web3Function.ts:137]: I think you could just add: 
```
      log: ctxData. log,
```
on the `context` creation, and keep:
```
    const result = await this._onRun(context);
```

File: src/lib/Web3Function.ts [hunk 3]
```diff
@@ -143,9 +156,9 @@ export class Web3Function {
     return Web3Function.Instance;
   }
 
-  static onRun(
-    onRun: (ctx: Web3FunctionContext) => Promise<Web3FunctionResult>
-  ): void {
+  static onRun(onRun: baseRunHandler): void;
+  static onRun(onRun: eventRunHandler): void;
+  static onRun(onRun: any): void {
```
Comment (goums) [inline src/lib/Web3Function.ts:161]: wouldn't this work?

```suggestion
  static onRun(onRun: baseRunHandler | eventRunHandler): void {
```

File: src/web3-functions/event/index.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  Web3Function,
+  Web3FunctionEventContext,
+} from "@gelatonetwork/web3-functions-sdk";
+
+Web3Function.onRun(async (context: Web3FunctionEventContext) => {
+  const { log } = context;
+
+  if (log.blockNumber !== 9727562) {
+    throw new Error("Log file fetched invalid");
+  }
```
Comment (goums) [inline src/web3-functions/event/index.ts:11]: could you include a simple log parsing & decoding as this could serve for reference

**Review by goums — CHANGES_REQUESTED** [review]

File: src/lib/Web3Function.ts (hunk 2 above)
Comment (mkykadir) [reply src/lib/Web3Function.ts:137]: For DX, I wanted to enforce log is defined for event triggering functions

Comment (goums) [reply src/lib/Web3Function.ts:137]: I don't understand what is the difference in DX between the suggestion:
```
const context = {
  ...
  log: ctxData.log
}
const result = await this._onRun(context);
```
And your implementation?
as your condition is on `ctxData.log`

File: src/web3-functions/event/index.ts (hunk 4 above)
Comment (mkykadir) [reply src/web3-functions/event/index.ts:24]: added in 4ef70ea620aff6abd690767799efb05e6e0b6f76

**Review by goums — APPROVED** [review]

**Pull request merged**